jsf/subsystem/src/main/java/org/jboss/as/jsf/deployment/JSFMetadataProcessor.java

```diff
@@ -81,22 +85,32 @@ public void deploy(DeploymentPhaseContext phaseContext) throws DeploymentUnitPro
         }
         if (disallowDoctypeDecl != null) {
             // Add the disallowDoctypeDecl context param if it's not already present
-            setContextParameterIfAbsent(metaData.getMergedJBossWebMetaData(), DISALLOW_DOCTYPE_DECL, disallowDoctypeDecl.toString());
+            setContextParameterIfAbsent(webMetaData, DISALLOW_DOCTYPE_DECL, disallowDoctypeDecl.toString());
         }
         // Auto-disable lazy bean validation for distributable web application.
         // This can otherwise cause missing @PreDestroy events.
-        if (metaData.getMergedJBossWebMetaData().getDistributable() != null) {
+        if (webMetaData.getDistributable() != null) {
             String disabled = Boolean.toString(false);
-            if (!setContextParameterIfAbsent(metaData.getMergedJBossWebMetaData(), LAZY_BEAN_VALIDATION_PARAM, disabled).equals(disabled)) {
+            if (!setContextParameterIfAbsent(webMetaData, LAZY_BEAN_VALIDATION_PARAM, disabled).equals(disabled)) {
                 JSFLogger.ROOT_LOGGER.lazyBeanValidationEnabled();
             }
         }
         // Set a default buffer size as 1024 is too small
-        final JBossWebMetaData webMetaData = metaData.getMergedJBossWebMetaData();
         // First check the legacy facelets.BUFFER_SIZE property which is required for backwards compatibility
         if (!hasContextParam(webMetaData, "facelets.BUFFER_SIZE")) {
             // The legacy parameter has not been set, set a default buffer if the current parameter name has not been set.
-            setContextParameterIfAbsent(metaData.getMergedJBossWebMetaData(), ViewHandler.FACELETS_BUFFER_SIZE_PARAM_NAME, Integer.toString(defaultBufferSize));
+            setContextParameterIfAbsent(webMetaData, ViewHandler.FACELETS_BUFFER_SIZE_PARAM_NAME, Integer.toString(defaultBufferSize));
+        }
+        // If web application is distributable, disable counter-productive "distributable" logic in Mojarra
+        if (JsfVersionMarker.getVersion(deploymentUnit).equals(JsfVersionMarker.JSF_4_0) && (webMetaData.getDistributable() != null)) {
+            ListenerMetaData mojarraListenerMetaData = new ListenerMetaData();
+            mojarraListenerMetaData.setListenerClass(ConfigureListener.class.getName());
```

Comment:
@pferraro @jasondlee Does this work if the JSF impl is MyFaces?

Reply:
AIUI, if the JSF impl is provided by the application (e.g. to use MyFaces or anything else), the JsfVersionMarker.getVersion(deploymentUnit) will return JsfVersionMarker.WAR_BUNDLES_JSF_IMPL.

Reply:
@pferraro We provide integration of MyFaces outside of the war. The latest mechanism for that is https://github.com/wildfly-extras/wildfly-myfaces-feature-pack.

Gating on JsfVersionMarker.JSF_4_0 may cover that as well; I'm just too lazy to research how it works.

Reply:
Ah... in that case, this fix (as is) will be problematic as it explicitly attaches the bootstrap listener for Mojarra.
@jasondlee Is there any way to distinguish the intended Faces implementation from this DUP?

Reply:
After further inspection, I think this logic is OK, so long as the application selects their preferred JSF implementation via the `org.jboss.jbossfaces.JSF_CONFIG_NAME` servlet context init parameter.
However, if the user instead installed the myfaces feature pack and configured the jsf subsystem to use a specific slot, e.g. myfaces, it is unclear to me how this is meant to work since `JsfVersionMarker.getVersion(deploymentUnit)` will still return "Mojarra-4.0" according to the logic within JSFVersionProcessor.  That does not seem right.  
@jasondlee Can you clarify how this is supposed to work?

Reply:
@pferraro TBH, I'm not sure ATM. That logic pre-dates me, but I'll take a look and figure it out.

Reply:
OK - I think it is sufficient to verify the slot of the implementation module, since the myfaces feature pack uses same module name but with "myfaces" slot.
Please verify the new logic.

Reply:
@jasondlee FYI, https://github.com/eclipse-ee4j/mojarra/pull/5347 was merged and will allow us to simplify this logic, i.e. rather than register servlet context listeners, we can just add the requisite Mojarra-specific servlet context init parameter (which will be safely ignored by other implementations).